assessment/libs/dates.php

```diff
@@ -0,0 +1,143 @@
+function dates_addWeeks($date, $weeks,$fmt = "F j, Y"){
+	$datetime = date_create_from_Format($fmt,$date);
+	if ($datetime==false) {
+		echo "date_addWeeks::could not create date from string. The date is " . $date . " and the format is " . $fmt . ".";
+		return false;
+	}
+	$datetime->modify('+' . $weeks .' weeks');
+	return $datetime->format($fmt);
+}
+
+// function addMonths
+// adds a number of months to an existing date
+// uses $fmt as format of date inputs and outputs
+function dates_addMonths($date,$months,$fmt= "F j, Y"){
+	$datetime = date_create_from_Format($fmt,$date);
+	if ($datetime==false) {
+		echo "date_addMonths::could not create date from string. The date is " . $date . " and the format is " . $fmt . ".";
+		return false;
+	}
+	$datetime->modify('+' . $months .' months');
+	return $datetime->format($fmt);
+}
+
+// function addYears
+// adds a number of weeks to an existing date
+// uses $fmt as format of date inputs and outputs
+function dates_addYears($date, $years,$fmt = "F j, Y"){
+	$datetime = date_create_from_Format($fmt,$date);
+	if ($datetime==false) {
+		echo "date_addYears::could not create date from string. The date is " . $date . " and the format is " . $fmt . ".";
+		return false;
+	}
+	$datetime->modify('+' . $years .' Years');
+	return $datetime->format($fmt);
+}
+
+// function EOMonth
+// returns the last date of the month
+// from a given date. 
+// uses $fmt as format of date inputs and outputs
+function dates_EOMonth($date,$fmt="F j, Y") {
+	$datetime = date_create_from_Format($fmt,$date);
+	if ($datetime==false) {
+		echo "date_EOMonth::could not create date from string. The date is " . $date . " and the format is " . $fmt . ".";
+		return false;
+	}
+	$datetime->modify('last day of this month');
+	return $datetime->format($fmt);
+}
+
+// function BOMonth
+// returns the first date of the month
+// from a given date. 
+// uses $fmt as format of date inputs and outputs
+function dates_BOMonth($date,$fmt="F j, Y") {
+	$datetime = date_create_from_Format($fmt,$date);
+	if ($datetime==false) {
+		echo "date_BOMonth::could not create date from string. The date is " . $date . " and the format is " . $fmt . ".";
+		return false;
+	}
+	$datetime->modify('first day of this month');
+	return $datetime->format($fmt);
+}
+
+// function diffDays
+// returns the number of days between
+// date1 and date2. Positive if date2>date1
+// uses $fmt as format of date inputs and outputs
+function dates_diffDays($date1,$date2,$fmt="F j, Y") {
+	$dt1 = date_create_from_format($fmt,$date1);
+	if ($dt1==false) {
+		echo "date_diffDays::could not create date from string. The date is " . $date1 . " and the format is " . $fmt . ".";
+		return false;
+	}
+	$dt2 = date_create_from_format($fmt,$date2);
+	if ($dt2==false) {
+		echo "date_diffDays::could not create date from string. The date is " . $date2 . " and the format is " . $fmt . ".";
+		return false;
+	}
+
+	$interval = $dt1->diff($dt2);
+	return (int) $interval->format("%r%a");
```

Comment:
What happens if the difference is not a whole number of days?  Should this be rounded rather than cast to int?

Reply:
That's a great question and a use case I had not thought about.  As this is for financial applications like interest calculations, I will truncate first and then take the difference (and mention this in the documentation). One typically assumes that one receives a full day of interest irrespective of when during the day a deposit is made.

Reply:
Hi,

I just tested the functionality and found that the existing implementation was as I wanted.  The difference between today at 3:10pm and tomorrow at 3:10am is 1 day and the difference between today at 3:10pm and tomorrow at 5:30pm are both one. I updated the test question to test this specific functionality.